### DISCUSSION TITLE
libvips calling 'getpeername'???

### DISCUSSION THREAD

**Discussion by yxiong20** [libvips/libvips#2541] [Q&A]
Hi libvips community:

We recently start using libvips and running our app in nsjail, it reports a seccomp violation of making 'getpeername' syscall. I'm wondering why libvips uses anything socket related, does it internally use some Unix socket?

Thanks,

Yufeng

**Comment by jcupitt** [comment]
Hi @yxiong20,

No, nothing in libvips will use that directly, as far as I know. It's probably one of the dependencies, but I don't know which one.

Have you tried putting a breakpoint on it and getting a stack trace?

**Reply by yxiong20** [reply]
Thanks John, I haven't run a debugger for the past 10 years but surprisingly I still know how to do it :)

So here is the stack trace:

```
#0  getpeername () at ../sysdeps/unix/syscall-template.S:120
#1  0x00007ffff7e8f7e5 in g_log_writer_is_journald ()
    at /usr/lib/x86_64-linux-gnu/libglib-2.0.so.0
#2  0x00007ffff7e90a06 in g_log_writer_default ()
    at /usr/lib/x86_64-linux-gnu/libglib-2.0.so.0
#3  0x00007ffff7e8ec67 in g_log_structured_array ()
    at /usr/lib/x86_64-linux-gnu/libglib-2.0.so.0
#4  0x00007ffff7e8f0b5 in g_log_default_handler ()
    at /usr/lib/x86_64-linux-gnu/libglib-2.0.so.0
#5  0x00007ffff7e8f309 in g_logv ()
    at /usr/lib/x86_64-linux-gnu/libglib-2.0.so.0
#6  0x00007ffff7e8f59f in g_log ()
    at /usr/lib/x86_64-linux-gnu/libglib-2.0.so.0
#7  0x00007ffff79d938b in  () at /usr/lib/x86_64-linux-gnu/libvips.so.42
#8  0x00007ffff7aa9e71 in  () at /usr/lib/x86_64-linux-gnu/libvips.so.42
#9  0x00007ffff7aaaad5 in vips_region_fill ()
    at /usr/lib/x86_64-linux-gnu/libvips.so.42
#10 0x00007ffff7aae80d in vips_region_prepare ()
    at /usr/lib/x86_64-linux-gnu/libvips.so.42
#11 0x00007ffff79da33a in  () at /usr/lib/x86_64-linux-gnu/libvips.so.42
#12 0x00007ffff7aa9e71 in  () at /usr/lib/x86_64-linux-gnu/libvips.so.42
#13 0x00007ffff7aaaad5 in vips_region_fill ()
    at /usr/lib/x86_64-linux-gnu/libvips.so.42
--Type <RET> for more, q to quit, c to continue without paging--
#14 0x00007ffff7aae80d in vips_region_prepare ()
    at /usr/lib/x86_64-linux-gnu/libvips.so.42
#15 0x00007ffff79e4a4e in  () at /usr/lib/x86_64-linux-gnu/libvips.so.42
#16 0x00007ffff7aa9e71 in  () at /usr/lib/x86_64-linux-gnu/libvips.so.42
#17 0x00007ffff7aaaad5 in vips_region_fill ()
    at /usr/lib/x86_64-linux-gnu/libvips.so.42
#18 0x00007ffff7aae80d in vips_region_prepare ()
    at /usr/lib/x86_64-linux-gnu/libvips.so.42
#19 0x00007ffff7a9b8fa in  () at /usr/lib/x86_64-linux-gnu/libvips.so.42
#20 0x00007ffff7aa9e71 in  () at /usr/lib/x86_64-linux-gnu/libvips.so.42
#21 0x00007ffff7aab037 in  () at /usr/lib/x86_64-linux-gnu/libvips.so.42
#22 0x00007ffff7aaea7d in vips_region_prepare_to ()
    at /usr/lib/x86_64-linux-gnu/libvips.so.42
#23 0x00007ffff7aa3379 in  () at /usr/lib/x86_64-linux-gnu/libvips.so.42
#24 0x00007ffff7aaf40a in  () at /usr/lib/x86_64-linux-gnu/libvips.so.42
#25 0x00007ffff7aaf18d in  () at /usr/lib/x86_64-linux-gnu/libvips.so.42
#26 0x00007ffff7eb10bd in  () at /usr/lib/x86_64-linux-gnu/libglib-2.0.so.0
#27 0x00007ffff7e1cea7 in start_thread (arg=<optimized out>)
    at pthread_create.c:477
#28 0x00007ffff781fdef in clone ()
    at ../sysdeps/unix/sysv/linux/x86_64/clone.S:95
```

Apparently it came from glib which is a direct dependency of libvips, seems to be logging related, it seems log message goes to system journal through a UNIX socket, per https://github.com/GNOME/glib/blob/main/glib/gmessages.c#L2211.

If I run it without logging that should avoid the problem? How do I configure vips to run without system journal logging? Is it required?

**Reply by yxiong20** [reply]
Also from the code it looks like it would always be called for any JPEG image processing, but in my case it is called only for certain images, such as one identified with 'file' cmd as "JPEG image data, Exif standard: [TIFF image data...". And I'm wondering why.

**Reply by jcupitt** [reply]
What's the message? Or you could build with debug enabled (-g, don't set -O) and you'll get line numbers for the function calls.

**Reply by jcupitt** [reply]
(you can define the `VIPS_WARNING` env var to block warning messages, if this is a warning message)

**Reply by yxiong20** [reply]
I think in my case vips won't be able to process the image anyway, this is the type of messages I saw during the processing:

> VIPS-WARNING **: 13:30:13.123: error in tile 0 x 0
> VIPS-WARNING **: 13:30:13.142: error in tile 0 x 8
> vips error: VipsJpeg: Invalid SOS parameters for sequential JPEG
> VipsJpeg: out of order read at line 0
> GLib-CRITICAL **: 14:05:49.467: g_hash_table_lookup: assertion 'hash_table != NULL' failed

The last line is from a different run that shows only when I compile my source with -g; I haven't tried recompile libvips with -g yet, I just linked with precompiled .so.

**Reply by jcupitt** [reply]
It sounds like a bad JPEG image. I could have a look if you can upload it somewhere.

**Reply by yxiong20** [reply]
It seems similar to the issue discussed in https://github.com/libvips/libvips/issues/2290.

**Reply by yxiong20** [reply]
> It sounds like a bad JPEG image. I could have a look if you can upload it somewhere.

The example I used is more than 10MB and cannot be uploaded here.

**Comment by yxiong20** [comment]
I tried another smaller file, it got a different type of error, but the output was generated.

> VIPS-WARNING **: 14:55:01.322: VipsJpeg: large XMP not saved

So sounds like any VIPS warning will get the message logged to system journal which causes the getpeername call I mentioned above because that is done via a Unix socket.

**Reply by jcupitt** [reply]
Have you tried setting `VIPS_WARNING`? It should suppress warning messages.

**Comment by yxiong20** [comment]
I did, it suppressed warning, but seems for more severe error type msg it still spills out, but that may be a good thing as it is for the case where the output cannot be generated anyway.

**Reply by jcupitt** [reply]
You can register your own log function with g_log_set_handler() and catch error messages.
